src/api/exchange/DepositApiMock.ts

```diff
@@ -104,7 +104,9 @@ export class DepositApiMock implements DepositApi {
     pendingDeposits.batchId = currentBatchId
 
     // mock transfer tokens from user's mock `wallet`
-    await this._erc20Api.transfer(tokenAddress, userAddress, this.getContractAddress(), amount)
+    await this._erc20Api.transfer(tokenAddress, userAddress, this.getContractAddress(), amount, {
+      senderAddress: this.getContractAddress(),
```

Comment:
As said in the comment, I understand we don't need this, the `fromAddress` (userAddress here) makes the `from`.
In addition, the sender cannot be the contract address :)

Reply:
@anxolin Wait, who should be the sender then?

Reply:
The user is depositing tokens in the exchange, so they are tokens that are deducted from the user's balanace and goes to the contract balance:
I don't see why we need the sender:
```js
await this._erc20Api.transfer(tokenAddress, userAddress, this.getContractAddress(), amount)
```

Reply:
Let's go through the process to check if we are on the same line:

1. User enables token:
  a. A `approve` tx is created with `spender` == contract address, `from` == user address, `token` == token address, `amount` == max amount.
1. User deposits token:
  a. A `transferFrom` tx is executed from the ERC20 contract (?) with `sender` == (user or contract?), `from` == user address, `to` == contract address
  b. Depending on who was the `sender` in the previous step, reduce the allowance.
1. Withdraw request -- not relevant
1. User withdraws
  a. A `transferFrom` tx is executed from the ERC20 contract (?) with `sender` == (user or contract?), `from` == contract address, `to` == user address
  b. Depending on who was the `sender` in the previous step, reduce the allowance.

Now that I was writing this I got confusing regarding where the tx originates from and at what point the allowance will be deduced.

@anxolin can you point where I got this wrong?

Reply:
I think the confusion was what I explained in the other thread. In case of doubt check the contract, cause the mock tries to mimic it, so as you see is what you say, it's a `transferFrom(user, contract, amount)`  where the sender is in fact the contract. We won't be needing to call this function in the app, but in the mock implementation makes sense.

![image](https://user-images.githubusercontent.com/2352112/66160127-c5515480-e629-11e9-88d0-507855ead372.png)